builtin/logical/ad/config/manager.go

```diff
@@ -22,8 +22,19 @@ const (
 )
 
 type Reader interface {
-	// Config returns a non-pointer to indicate the the engineConf returned is immutable
-	Config(ctx context.Context, storage logical.Storage) (ImmutableEngineConf, error)
+	// Config returns the current *EngineConf.
+	// This shouldn't be retained in memory, but rather should be pulled on the fly every time,
+	// in case the config changes.
+	//   - If error is nil, *EngineConf is not nil.
+	//   - Returns an UnsetError if the config is currently unset by the user - none exists.
+	//   - Returns other errors if storage is inaccessible, etc.
+	Config(ctx context.Context, storage logical.Storage) (*EngineConf, error)
+}
+
+type UnsetError struct{}
```

Comment:
Oh, because it's a totally different error. I think `ErrReadOnly` is used for when people try to write to something that they can't write to.

The `UnsetError` will look like this to anyone using it: `config.UnsetError`, which is meant to convey that the user doesn't currently have any config set. I expect that it'll be an edge case that anybody would be trying to use the AD secrets backend without an active config that the operator set up, but it technically could happen so must be accounted for.

I found that other back ends tend to return `nil` in that situation, not accompanied by any error. That scares me a little because if there's no error, I always assume it's safe to touch the other thing(s) that came out of a method, and I don't want to cause panics for anybody using the code who doesn't know that can happen.

Strongly typed errors are often used to flag certain situations, like `io.EOF` for instance, so I thought it best to return an error if the pointer would be nil, and I figure most of the time the caller will want to return because they probably need a config for whatever they're doing. But if they really need to know whether it's a "real" error or it was just unset, they _can_ test it to find out.

Reply:
I didn't mean that exact error, it was just an example of using the standard `errors.New(...)`.  But I see what you're going for here. If an empty config is not really an error, but just a state, then a nil response doesn't seem that bad. You can still return an error parameter too, which would just be nil as well in the empty config case. It eliminates the new type (and dependency if you need to check that type), and is a little flatter:
```
// current
config, err = configReader.Config(ctx, storage)
if err != nil {
	_, ok := err.(*UnsetError)
	if !ok {
		t.FailNow()
	}
	// Handle empty config
}

// proposed
config, err = configReader.Config(ctx, storage)
if err != nil {
	t.FailNow()
}
if config == nil {
	// Handle empty config
}
```

Reply:
Kk, will update.